In [1]:
import pandas as pd
import numpy as np

## ĆWICZENIE 2

#### 1. Pobierz zbiór danych osób, które uczestniczyły w rejsie statkiem Titanic

In [2]:
df = pd.read_csv("C://Users/48510/Desktop/ML/dataML.csv")
df.head(20)

pclass  survived                                               name  \
0        1         1                      Allen, Miss. Elisabeth Walton   
1        1         1                     Allison, Master. Hudson Trevor   
2        1         0                       Allison, Miss. Helen Loraine   
3        1         0               Allison, Mr. Hudson Joshua Creighton   
4        1         0    Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
5        1         1                                Anderson, Mr. Harry   
6        1         1                  Andrews, Miss. Kornelia Theodosia   
7        1         0                             Andrews, Mr. Thomas Jr   
8        1         1      Appleton, Mrs. Edward Dale (Charlotte Lamson)   
9        1         0                            Artagaveytia, Mr. Ramon   
10       1         0                             Astor, Col. John Jacob   
11       1         1  Astor, Mrs. John Jacob (Madeleine Talmadge Force)   
12       1         1                      Aubart, Mme. Leontine Pauline   
13       1         1                       Barber, Miss. Ellen 'Nellie'   
14       1         1               Barkworth, Mr. Algernon Henry Wilson   
15       1         0                                Baumann, Mr. John D   
16       1         0                           Baxter, Mr. Quigg Edmond   
17       1         1    Baxter, Mrs. James (Helene DeLaudeniere Chaput)   
18       1         1                              Bazzani, Miss. Albina   
19       1         0                               Beattie, Mr. Thomson   

       sex     age  sibsp  parch    ticket      fare    cabin embarked boat  \
0   female      29      0      0     24160  211.3375       B5        S    2   
1     male  0.9167      1      2    113781    151.55  C22 C26        S   11   
2   female       2      1      2    113781    151.55  C22 C26        S    ?   
3     male      30      1      2    113781    151.55  C22 C26        S    ?   
4   female      25      1      2    113781    151.55  C22 C26        S    ?   
5     male      48      0      0     19952     26.55      E12        S    3   
6   female      63      1      0     13502   77.9583       D7        S   10   
7     male      39      0      0    112050         0      A36        S    ?   
8   female      53      2      0     11769   51.4792     C101        S    D   
9     male      71      0      0  PC 17609   49.5042        ?        C    ?   
10    male      47      1      0  PC 17757   227.525  C62 C64        C    ?   
11  female      18      1      0  PC 17757   227.525  C62 C64        C    4   
12  female      24      0      0  PC 17477      69.3      B35        C    9   
13  female      26      0      0     19877     78.85        ?        S    6   
14    male      80      0      0     27042        30      A23        S    B   
15    male       ?      0      0  PC 17318    25.925        ?        S    ?   
16    male      24      0      1  PC 17558  247.5208  B58 B60        C    ?   
17  female      50      0      1  PC 17558  247.5208  B58 B60        C    6   
18  female      32      0      0     11813   76.2917      D15        C    8   
19    male      36      0      0     13050   75.2417       C6        C    A   

   body                        home.dest  
0     ?                     St Louis, MO  
1     ?  Montreal, PQ / Chesterville, ON  
2     ?  Montreal, PQ / Chesterville, ON  
3   135  Montreal, PQ / Chesterville, ON  
4     ?  Montreal, PQ / Chesterville, ON  
5     ?                     New York, NY  
6     ?                       Hudson, NY  
7     ?                      Belfast, NI  
8     ?              Bayside, Queens, NY  
9    22              Montevideo, Uruguay  
10  124                     New York, NY  
11    ?                     New York, NY  
12    ?                    Paris, France  
13    ?                                ?  
14    ?                    Hessle, Yorks  
15    ?                     New York, NY  
16    ?                     Montreal,

#### 2. Dokonaj wstępnej inspekcji pierwszych 20 wierszy. Co widzisz?
##### po w stępnej analizie pierwszych 20 wierszy można zauważyć, że w zbiorze danych są brakujące wartości - ?

#### 3. Sprawdź ile jest wartości brakujących Czy zbiór danych należy zmodyfikować przed użyciem w/w funkcji? Jeśli tak to jak? → wystarczy jedna linia kodu , zapisz swoje obserwację i skomentuj % brakującychdanych w zależności od danej zmiennej.    
#####      Odp.Tak, trzeba wypełnić puste wartości wartościami NaN

In [3]:
df=df.replace("?",np.NaN)
missing_values = df.isnull().sum()
print(missing_values)

pclass          0
survived        0
name            0
sex             0
age           263
sibsp           0
parch           0
ticket          0
fare            1
cabin        1014
embarked        2
boat          823
body         1188
home.dest     564
dtype: int64


In [4]:
missing_values_mean = df.isnull().mean()
print(missing_values_mean)

pclass       0.000000
survived     0.000000
name         0.000000
sex          0.000000
age          0.200917
sibsp        0.000000
parch        0.000000
ticket       0.000000
fare         0.000764
cabin        0.774637
embarked     0.001528
boat         0.628724
body         0.907563
home.dest    0.430863
dtype: float64


##### braki danych występują w kolumnach: age 20%, cabin 77%, boat 63%, body 91%, home.dest 43% i są to poważne braki danych, natomiast mniejsze braki danych <1% są w kolumnach fare i embarked

### 4. z jakim rodzajem braków danych masz do czynienia – całkowicie przypadkowe (MCAR), przypadkowe (MAR) czy nie przypadkowe (MNAR)?

In [5]:
null_count=df.isnull().sum().sum()
print(null_count)

3855


In [6]:
df.shape

(1309, 14)

In [7]:
nulls_percentage=null_count/(1309*14)*100
print(nulls_percentage)

21.035687002073555


In [8]:
df['home.destNull']= np.where(df['home.dest'].isnull(), 1, 0)
df.groupby(['survived'])['home.destNull'].mean()

survived
0    0.508035
1    0.306000
Name: home.destNull, dtype: float64

In [9]:
df['home.dest'].isna().groupby(df['survived']).mean()

survived
0    0.508035
1    0.306000
Name: home.dest, dtype: float64

In [10]:
df[['fare','embarked','age','cabin','boat','body','home.dest']].isna().groupby(df['survived']).mean()

fare  embarked       age     cabin      boat      body  \
survived                                                               
0         0.001236     0.000  0.234858  0.873918  0.988875  0.850433   
1         0.000000     0.004  0.146000  0.614000  0.046000  1.000000   

          home.dest  
survived             
0          0.508035  
1          0.306000

##### braków danych w całym zbiorze jest ok 21%.Według mnie braki danych nie są przypadkowe - są to braki typu MNAR – ich ilość wskazuje na to że nie możemy ich usunąć. Niektóre braki danych są związane z liczbą ocalałych osób, widać to np. w kolumnie cabin – w większości przypadków z brakiem przeżycia wiąże się brak informacji o kabinie. Duży wpływ na NaN ma wiek zbioru, dane są bardzo stare – przez taki okres czasu niektóre dane mogły zostać zgubione/utracone, albo w ogóle nie zostały zebrane po katastrofie. 

### 5. Po powyższej analizie odpowiedz na pytanie w jaki sposób należy postąpić z brakującymi wartościami. 

##### Braki w kolumnach fare,embarked można uznac za całkowicie przypadkowe ponieważ są <5% więc można usunąć te rekordy. Braki w kolumnach age możemy uzupełnić wartością średnią lub medianą (w zalezności od rozkładu zmiennych). Dla cabin, boat, body, home.dest stworzyłabym nową kategorię "missing"

## ĆWICZENIE 3

### 1. Sprawdź liczebność poszczególnych etykiet dla danych zmiennych jakościowych

In [11]:
print('Liczba etykiet zmiennej name: {}'.format(len(df.name.unique())))
print('Liczba etykiet zmiennej pclass: {}'.format(len(df.pclass.unique())))
print('Liczba etykiet zmiennej sibsp: {}'.format(len(df.sibsp.unique())))
print('Liczba etykiet zmiennej sex: {}'.format(len(df.sex.unique())))
print('Liczba etykiet zmiennej parch: {}'.format(len(df.parch.unique())))
print('Liczba etykiet zmiennej embarked: {}'.format(len(df.embarked.unique())))
print('Liczba etykiet zmiennej cabin: {}'.format(len(df.cabin.unique())))
print('Liczba etykiet zmiennej boat: {}'.format(len(df.boat.unique())))
print('Liczba etykiet zmiennej home.dest: {}'.format(len(df['home.dest'].unique())))

Liczba etykiet zmiennej name: 1307
Liczba etykiet zmiennej pclass: 3
Liczba etykiet zmiennej sibsp: 7
Liczba etykiet zmiennej sex: 2
Liczba etykiet zmiennej parch: 8
Liczba etykiet zmiennej embarked: 4
Liczba etykiet zmiennej cabin: 187
Liczba etykiet zmiennej boat: 28
Liczba etykiet zmiennej home.dest: 370


### 2. Wyświetl z użyciem funkcji print liczbę wszystkich pasażerów.

In [12]:
print('Liczba wszystkich pasażerów: {}'.format(len(df)))

Liczba wszystkich pasażerów: 1309


### 4. Skomentuj wyniki otrzymane w punkcie 1 i 2. Podziel zmienne ze względu na dużą i małą moc zbioru (kardynalność).

##### Niektóre zmienne mają tylko kilka etykiet a inne bardzo dużo, mogą być unikalne nawet dla każdego pasażera(home.dest)
##### Zmienne z dużą mocą zbioru:name,cabin, home.dest
##### Zmienne z małą mocą zbioru: pozostałe 

### 5. Sprawdź ile unikalnych etykiet ma zmienna mówiąca o kabinie danego pasażera. Użyj takiej funkcji, która zwraca wynik w postaci NumPy array. Jaka jest różnica pomiędzy NumPy array a pandas DataFrame?

In [13]:
print('Liczba unikalnych etykiet zmiennej mówiącej o kabinie danego pasażera: {}'.format(len(df['cabin'].unique())))

Liczba unikalnych etykiet zmiennej mówiącej o kabinie danego pasażera: 187


In [14]:
print(type(df['cabin'].unique()))

<class 'numpy.ndarray'>


In [15]:
df['cabin'].unique()

array(['B5', 'C22 C26', 'E12', 'D7', 'A36', 'C101', nan, 'C62 C64', 'B35',
       'A23', 'B58 B60', 'D15', 'C6', 'D35', 'C148', 'C97', 'B49', 'C99',
       'C52', 'T', 'A31', 'C7', 'C103', 'D22', 'E33', 'A21', 'B10', 'B4',
       'E40', 'B38', 'E24', 'B51 B53 B55', 'B96 B98', 'C46', 'E31', 'E8',
       'B61', 'B77', 'A9', 'C89', 'A14', 'E58', 'E49', 'E52', 'E45',
       'B22', 'B26', 'C85', 'E17', 'B71', 'B20', 'A34', 'C86', 'A16',
       'A20', 'A18', 'C54', 'C45', 'D20', 'A29', 'C95', 'E25', 'C111',
       'C23 C25 C27', 'E36', 'D34', 'D40', 'B39', 'B41', 'B102', 'C123',
       'E63', 'C130', 'B86', 'C92', 'A5', 'C51', 'B42', 'C91', 'C125',
       'D10 D12', 'B82 B84', 'E50', 'D33', 'C83', 'B94', 'D49', 'D45',
       'B69', 'B11', 'E46', 'C39', 'B18', 'D11', 'C93', 'B28', 'C49',
       'B52 B54 B56', 'E60', 'C132', 'B37', 'D21', 'D19', 'C124', 'D17',
       'B101', 'D28', 'D6', 'D9', 'B80', 'C106', 'B79', 'C47', 'D30',
       'C90', 'E38', 'C78', 'C30', 'C118', 'D36', 'D48', 'D47', '

##### DATAFRAME VS NUMPY ARRAY : Pandas Data Frame zawiera w sobie NumPy array, tablice Numpy mogą być wielowymiarowe, podczas gdy DataFrame może być tylko dwuwymiarowe, tablice Numpy są szczególnie używane, gdy konieczne jest wykonanie złożonych obliczeń naukowych, podczas gdy ramki DataFrames są używane głównie do wstępnego przetwarzania danych. 

### 6. Zredukuj liczbę cech dla zmiennej opisującej kabiny poprzez zastąpienie obecnych etykiet w formacie LL11 do etykiet zawierających tylko pierwszą literę.

In [16]:
df['CabinReduced']=df['cabin'].astype(str).str[0]

In [17]:
df[['cabin','CabinReduced']].head(20)

cabin CabinReduced
0        B5            B
1   C22 C26            C
2   C22 C26            C
3   C22 C26            C
4   C22 C26            C
5       E12            E
6        D7            D
7       A36            A
8      C101            C
9       NaN            n
10  C62 C64            C
11  C62 C64            C
12      B35            B
13      NaN            n
14      A23            A
15      NaN            n
16  B58 B60            B
17  B58 B60            B
18      D15            D
19       C6            C

### 7. Wyświetl (jak w punkcie 1) liczbę etykiet dla zmiennych z punkcie 5. O ile procent zredukowano kardynalność zbioru zmiennej opisującej kabiny?

In [18]:
print('Liczba etykiet zmiennej cabin: {}'.format(len(df['cabin'].unique())))
print('Liczba etykiet zmiennej CabinReduced: {}'.format(len(df['CabinReduced'].unique())))
reduced=(1-len(df.CabinReduced.unique())/len(df.cabin.unique()))*100
print('kardynalność zbioru zmiennej opisującej kabiny zredukowano o: {} %'.format(round(reduced,4)))

Liczba etykiet zmiennej cabin: 187
Liczba etykiet zmiennej CabinReduced: 9
kardynalność zbioru zmiennej opisującej kabiny zredukowano o: 95.1872 %


### 8. Uzasadnij dlaczego dokonujesz redukcji akurat tej zmiennej. Jak to wpływa na przyszłe analizy. Czy powoduje jakieś negatywne skutki?

#### Dokonuję redukcji tej zmiennej, ponieważ to znacząco ułatwi zadanie modelu ML – analiza będzie szybsza i łatwiejsza, ponieważ model będzie miał tylko 9 etykiet, a nie jak przed redukcją-187.  Negatywnym skutkiem takiej redukcji może być utrata części informacji o dokładnym położeniu kabiny, ponieważ po redukcji otrzymujemy informację tylko o sektorze/klasie kabiny. 

In [19]:
df.to_csv(r'C://Users/48510/Desktop/ML/datalab2.csv')